### DISCUSSION TITLE
Pagination with filters parameter ?

### DISCUSSION THREAD

**Discussion by plzapsys** [meysamhadeli/shop-golang-microservices#90] [Q&A] [answered]
Hi,
I'm trying to use pagination in combination with “filters” parameter like this: 
url "/api/v1/ou?size=2&page=1&filters=[field: "xxx", value: "yyy" , comparison: "zzz"]”

Could you please give me an example how to use pagination with “filters” parameter?
url  /api/v1/ou?size=2&page=1&filters=?

Thank You in advance!

**Comment by meysamhadeli** [comment]
> Hi, I'm trying to use pagination in combination with “filters” parameter like this: url "/api/v1/ou?size=2&page=1&filters=[field: "xxx", value: "yyy" , comparison: "zzz"]”
> 
> Could you please give me an example how to use pagination with “filters” parameter? url /api/v1/ou?size=2&page=1&filters=?
> 
> Thank You in advance!

Hi, I used the pagination in this part of code, and you can debug it for more clarifying how it works:
https://github.com/meysamhadeli/shop-golang-microservices/blob/main/internal/services/product_service/product/data/repositories/pg_product_repository.go#L30

And the request should be like this: 
http://your-api-endpoint?size=10&page=1&orderBy=created_at&filters[0][field]=name&filters[0][value]=John&filters[0][comparison]=eq&filters[1][field]=age&filters[1][value]=30&filters[1][comparison]=gte

**Reply by plzapsys** [reply]
Hi,
Thank You for your quick reply. I come from PHP world and I'm beginner in GO language. 
I have a PHP project that I want to rewrite to GO. 
Long thought about what technologies and architecture to choose. After I saw Your project and approach I had no hesitation that it is the the best solution for me.

After debugging in the following sequence:

1. Request to => URL
  => api/v1/ou?size=2&page=1&filters[0][field]=oul&filters[0][value]=GDPBZN&filters[0][comparison]=contains

2. https://github.com/meysamhadeli/shop-golang-microservices/blob/main/internal/services/product_service/product/features/getting_products/v1/endpoints/get_products_endpoint.go#L34

func getAllProducts(validator *validator.Validate, log logger.ILogger, ctx context.Context) echo.HandlerFunc {
	return func(c echo.Context) error {
		listQuery, err := utils.GetListQueryFromCtx(c) 


3. https://github.com/meysamhadeli/shop-golang-microservices/blob/main/internal/pkg/utils/pagination.go#L76

func GetListQueryFromCtx(c echo.Context) (*ListQuery, error) {
	q := &ListQuery{}
	var page, size, orderBy string
	err := echo.QueryParamsBinder(c).
		CustomFunc("filters", func(values []string) []error {
			for _, v := range values {
				if v == "" {
					continue
				}
				f := &FilterModel{}
				if err := c.Bind(f); err != nil {
					return []error{err}
				}
				q.Filters = append(q.Filters, f)
			}
			return nil
		}).
		String("size", &size).
		String("page", &page).
		String("orderBy", &orderBy).
		BindError() // returns first binding error

	
4. Result from the func GetListQueryFromCtx is: 
 => {"size":2,"page":1} 

At the end the "filters" parameter is mising in the result. I suppose that it should be:
 {"size":2,"page":1,"filters":[{"field":"oul","value":"GDPBZN","comparison":"equals"}]}
 
 Am I right or I'm missing something?

**Reply by meysamhadeli** [reply, marked as answer]
I didn't test the filter functionality before, and I port this filter functionality from .net to go. It seems something is wrong. Is it possible for you to fix this problem by a unit test? Unfortunately, currently I don't have time right on to investigate the problem.

**Reply by plzapsys** [reply]
Hi,
Unfortunately, I haven't found a way to get the filters parameter from a url as an array in Go.
I made a few corrections that work for me. I'm not sure if this is the best solution to the problem.
I have uploaded only the files in which I have made corrections, they can be seen here
https://github.com/plzapsys/go-pkg/tree/main/internal/pkg

**Reply by meysamhadeli** [reply]
Thanks for sharing. If I find a better solution, I will add it to project :)
